File: v1/src/main/java/com/api/v1/person/services/PersonRegistrationServiceImpl.java
package com.api.v1.person.services;

import com.api.v1.person.domain.Person;
import com.api.v1.person.domain.PersonRepository;
import com.api.v1.person.dtos.PersonRegistrationRequestDto;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;

@Service
class PersonRegistrationServiceImpl implements PersonRegistrationService {

    @Autowired
    private PersonRepository personRepository;

    @Override
    public Mono<Person> register(@Valid PersonRegistrationRequestDto requestDto) {
        return personRepository.save(Person.of(requestDto));
    }

}

File: v2/src/main/kotlin/com/api/v2/persons/services/PersonRegistrationServiceImpl.kt
package com.api.v2.persons.services

import com.api.v2.persons.domain.Person
import com.api.v2.persons.domain.PersonRepository
import com.api.v2.persons.dtos.PersonRegistrationDto
import com.api.v2.persons.exceptions.DuplicatedSsnException
import jakarta.validation.Valid
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
private class PersonRegistrationServiceImpl: PersonRegistrationService {

    @Autowired
    lateinit var personRepository: PersonRepository

    override suspend fun register(requestDto: @Valid PersonRegistrationDto): Person {
        return withContext(Dispatchers.IO) {
            isSsnDuplicated(requestDto.ssn)
            val person = Person.of(requestDto)
            personRepository.save(person)
        }
    }

    private suspend fun isSsnDuplicated(ssn: String) {
        if (personRepository.findBySsn(ssn) != null) {
            throw DuplicatedSsnException()
        }
    }

}